crates/starknet-devnet-server/src/api/json_rpc/endpoints_ws.rs

```diff
@@ -97,15 +114,96 @@ impl JsonRpcHandler {
         // Yes, only if block_id != latest/pending (handled above)
+        let starknet = self.api.starknet.lock().await;
         for block_n in query_block_number..=latest_block_number {
             let old_block = starknet
                 .get_block(&BlockId::Number(block_n))
                 .map_err(ApiError::StarknetDevnetError)?;
 
-            let notification = SubscriptionNotification::NewHeadsNotification(old_block.into());
+            let old_header = old_block.into();
+            let notification = SubscriptionNotification::NewHeads(Box::new(old_header));
             socket_context.notify(subscription_id, notification).await;
         }
 
         Ok(())
     }
+
+    /// Based on block generation mode and specified block ID, decide on subscription's sensitivity:
+    /// notify of changes in pending or latest block
+    fn get_subscription_tag(&self, block_id: BlockId) -> BlockTag {
+        match self.starknet_config.block_generation_on {
+            BlockGenerationOn::Transaction => BlockTag::Latest,
+            BlockGenerationOn::Demand | BlockGenerationOn::Interval(_) => match block_id {
+                BlockId::Tag(tag) => tag,
+                BlockId::Hash(_) | BlockId::Number(_) => BlockTag::Pending,
+            },
+        }
+    }
+
+    async fn subscribe_tx_status(
+        &self,
+        transaction_block_input: TransactionBlockInput,
+        rpc_request_id: Id,
+        socket_id: SocketId,
+    ) -> Result<(), ApiError> {
+        let TransactionBlockInput { transaction_hash, block } = transaction_block_input;
+
+        let query_block_id = if let Some(block_id) = block {
+            block_id.0
+        } else {
+            // if no block ID input, this eventually just subscribes the user to new blocks
+            BlockId::Tag(BlockTag::Latest)
+        };
+
+        let (query_block_number, latest_block_number) =
+            self.convert_to_block_number_range(query_block_id).await?;
+
+        // perform the actual subscription
+        let mut sockets = self.api.sockets.lock().await;
+        let socket_context = sockets.get_mut(&socket_id).ok_or(ApiError::StarknetDevnetError(
+            Error::UnexpectedInternalError { msg: format!("Unregistered socket ID: {socket_id}") },
+        ))?;
+
+        // TODO if tx present, but in a block before the one specified, no point in subscribing -
+        // its status shall never change (unless considering block abortion). It would make
+        // sense to just add a ReorgSubscription
+        let subscription = Subscription::TransactionStatus {
+            tag: self.get_subscription_tag(query_block_id),
+            transaction_hash,
+        };
+        let subscription_id = socket_context.subscribe(rpc_request_id, subscription).await;
+
+        let starknet = self.api.starknet.lock().await;
+
+        if let (Ok(receipt), Ok(status)) = (
+            starknet.get_transaction_receipt_by_hash(&transaction_hash),
```

Comment:
why there is a second call to get the execution and finality status, when the properties are in the receipt

Reply:
If I remember correctly, I would have to reimplement some logic that is already implemented in the call to execution and finality status.

Reply:
But looking at it again, it's not that much logic.

Reply:
Addressed in https://github.com/0xSpaceShard/starknet-devnet-rs/pull/646/commits/de7709335088edf3643318e4c56e155f04c84bb0